pysteps/nowcasts/anvil.py

```diff
@@ -0,0 +1,370 @@
+                               allow_nonfinite_values=True, **extrap_kwargs)[-1]
+
+    for i in range(vil.shape[0] - 1):
+        if not DASK_IMPORTED or num_workers == 1:
+            vil[i, :, :] = worker(vil, i)[1]
+        else:
+            res.append(dask.delayed(worker)(vil, i))
+
+    if DASK_IMPORTED and num_workers > 1:
+        num_workers_ = len(res) if num_workers > len(res) else num_workers
+        vil_e = dask.compute(*res, num_workers=num_workers_)
+        for i in range(len(vil_e)):
+            vil[vil_e[i][0], :] = vil_e[i][1]
+
+    mask = np.isfinite(vil[0, :])
+    for i in range(1, vil.shape[0]):
+        mask = np.logical_and(mask, np.isfinite(vil[i, :]))
+
+    bp_filter_method = cascade.get_method("gaussian")
+    bp_filter = bp_filter_method((m, n), n_cascade_levels, **filter_kwargs)
+
+    fft = utils.get_method(fft_method, shape=vil.shape[1:],
+                           n_threads=num_workers)
+
+    decomp_method, recomp_method = cascade.get_method("fft")
+
+    vil_dec = np.empty((n_cascade_levels, vil.shape[0], m, n))
+    for i in range(vil.shape[0]):
+        vil_ = vil[i, :].copy()
+        vil_[~np.isfinite(vil_)] = 0.0
+        vil_dec_i = decomp_method(vil_, bp_filter, fft_method=fft)
+        for j in range(n_cascade_levels):
+            vil_dec[j, i, :] = vil_dec_i["cascade_levels"][j, :]
+
+    gamma = np.empty((n_cascade_levels, ar_order, m, n))
+    for i in range(n_cascade_levels):
+        vil_diff = np.diff(vil_dec[i, :], axis=0)
+        vil_diff[~np.isfinite(vil_diff)] = 0.0
+        for j in range(ar_order):
+            gamma[i, j, :] = _moving_window_corrcoef(vil_diff[-1, :],
+                                                     vil_diff[-(j+2), :],
+                                                     ar_window_radius)
+
+    if ar_order == 2:
+        for i in range(n_cascade_levels):
+            gamma[i, 1, :] = autoregression.adjust_lag2_corrcoef2(gamma[i, 0, :],
+                                                                  gamma[i, 1, :])
+
+    phi = []
+    for i in range(n_cascade_levels):
+        if ar_order > 2:
+            phi_ = autoregression.estimate_ar_params_yw_localized(gamma[i, :], d=1)
+        elif ar_order == 2:
+            phi_ = _estimate_ar2_params(gamma[i, :])
+        else:
+            pass
+        phi.append(phi_)
+
+    vil_dec = vil_dec[:, -3:, :]
+
+    if measure_time:
+        init_time = time.time() - starttime_init
+
+    print("Starting nowcast computation.")
+
+    if measure_time:
+        starttime_mainloop = time.time()
+
+    r_f = []
+    dp = None
+    for t in range(n_timesteps):
+        print("Computing nowcast for time step %d... " % (t + 1), end="")
+        sys.stdout.flush()
+        
+        if measure_time:
+            starttime = time.time()
+
+        for i in range(n_cascade_levels):
+            vil_dec[i, :] = autoregression.iterate_ar_model(vil_dec[i, :], phi[i])
```

Comment:
calling `anvil.forecast` with `ar_order=3` fails here with the following error message:

```
Computing ANVIL nowcast:
------------------------

Inputs:
-------
input dimensions: 320x355

Methods:
--------
extrapolation:   semilagrangian
FFT:             numpy

Parameters:
-----------
number of time steps:        3
parallel threads:            1
number of cascade levels:    8
order of the ARI(p,1) model: 3
ARI(p,1) window radius:      50
R(VIL) window radius:        5
Starting nowcast computation.
Computing nowcast for time step 1... Traceback (most recent call last):
  File "pysteps/tests/test_nowcasts_anvil.py", line 73, in <module>
    test_anvil_rainrate(**dict((x, y) for x, y in test_args))
  File "pysteps/tests/test_nowcasts_anvil.py", line 58, in test_anvil_rainrate
    ar_window_radius=ar_window_radius,
  File "/users/ned/pySTEPS/pysteps/pysteps/nowcasts/anvil.py", line 243, in forecast
    vil_dec[i, :] = autoregression.iterate_ar_model(vil_dec[i, :], phi[i])
  File "/users/ned/pySTEPS/pysteps/pysteps/timeseries/autoregression.py", line 975, in iterate_ar_model
    raise ValueError("dimension mismatch between x and phi: x.shape[0]=%d, len(phi)=%d" % (x.shape[0], len(phi)))
ValueError: dimension mismatch between x and phi: x.shape[0]=3, len(phi)=5
```

Reply:
> calling `anvil.forecast` with `ar_order=3` fails here with the following error message:
> 
> ```
> Computing ANVIL nowcast:
> ------------------------
> 
> Inputs:
> -------
> input dimensions: 320x355
> 
> Methods:
> --------
> extrapolation:   semilagrangian
> FFT:             numpy
> 
> Parameters:
> -----------
> number of time steps:        3
> parallel threads:            1
> number of cascade levels:    8
> order of the ARI(p,1) model: 3
> ARI(p,1) window radius:      50
> R(VIL) window radius:        5
> Starting nowcast computation.
> Computing nowcast for time step 1... Traceback (most recent call last):
>   File "pysteps/tests/test_nowcasts_anvil.py", line 73, in <module>
>     test_anvil_rainrate(**dict((x, y) for x, y in test_args))
>   File "pysteps/tests/test_nowcasts_anvil.py", line 58, in test_anvil_rainrate
>     ar_window_radius=ar_window_radius,
>   File "/users/ned/pySTEPS/pysteps/pysteps/nowcasts/anvil.py", line 243, in forecast
>     vil_dec[i, :] = autoregression.iterate_ar_model(vil_dec[i, :], phi[i])
>   File "/users/ned/pySTEPS/pysteps/pysteps/timeseries/autoregression.py", line 975, in iterate_ar_model
>     raise ValueError("dimension mismatch between x and phi: x.shape[0]=%d, len(phi)=%d" % (x.shape[0], len(phi)))
> ValueError: dimension mismatch between x and phi: x.shape[0]=3, len(phi)=5
> ```

It could be a good idea to disallow using a higher-order AR model (p>2). Not sure if the nowcasts are reliable because the stationarity of the AR process can only be guaranteed when p<=2.

Reply:
It stationarity is the problem, then it's probably enough to specify it somewhere in the docstrings.
Is the error above cause by that?

Reply:
Seems that is caused by line 241: `vil_dec = vil_dec[:, -3:, :]`, that keeps only the last 3 frames of the 4 passed in the tests. Does ANVIL always use only 3 frames?

Reply:
> If stationarity is the problem, maybe it is enough to specify it somewhere in the docstrings.
> But is this the cause for the error above?

The error is caused by a different problem. I'll fix that.

Reply:
Line 241 was indeed incorrect.